uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/test/java/org/uberfire/ext/wires/core/grids/client/widget/grid/renderers/grids/impl/BaseGridRendererSelectionLayerTest.java

```diff
@@ -0,0 +1,191 @@
+import org.uberfire.ext.wires.core.grids.client.widget.grid.renderers.grids.GridRenderer.RenderGridBoundaryCommand;
+import org.uberfire.ext.wires.core.grids.client.widget.grid.renderers.grids.GridRenderer.RenderHeaderBackgroundCommand;
+import org.uberfire.ext.wires.core.grids.client.widget.grid.renderers.grids.GridRenderer.RenderHeaderGridLinesCommand;
+import org.uberfire.ext.wires.core.grids.client.widget.grid.renderers.grids.GridRenderer.RenderSelectorCommand;
+import org.uberfire.ext.wires.core.grids.client.widget.grid.renderers.grids.GridRenderer.RendererCommand;
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.anyInt;
+import static org.mockito.Matchers.anyList;
+import static org.mockito.Matchers.anyObject;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.reset;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static org.uberfire.ext.wires.core.grids.client.widget.grid.impl.BaseGridWidgetRenderingTestUtils.ROW_HEIGHT;
+import static org.uberfire.ext.wires.core.grids.client.widget.grid.impl.BaseGridWidgetRenderingTestUtils.makeRenderingInformation;
+
+@WithClassesToStub({Text.class})
+@RunWith(LienzoMockitoTestRunner.class)
+public class BaseGridRendererSelectionLayerTest extends BaseGridRendererTest {
+
+    @Override
+    protected boolean isSelectionLayer() {
+        return true;
+    }
+
+    @Test
+    public void checkRenderSelector() {
+        final BaseGridRendererHelper.RenderingInformation ri = makeRenderingInformation(model,
+                                                                                        Arrays.asList(0d, ROW_HEIGHT, ROW_HEIGHT * 2));
+
+        final RendererCommand command = renderer.renderSelector(WIDTH,
+                                                                HEIGHT,
+                                                                ri);
+
+        assertNotNull(command);
+        assertRenderingCommands(Collections.singletonList(command),
+                                RenderSelectorCommand.class);
+
+        command.execute(rc);
+
+        verify(parent, never()).add(anyObject());
+    }
+
+    @Test
+    @SuppressWarnings("unchecked")
+    public void checkSelectedCells() {
+        this.model.selectCells(0, 0, 1, 1);
+        when(context.getMinVisibleRowIndex()).thenReturn(0);
+        when(context.getMaxVisibleRowIndex()).thenReturn(1);
+
+        renderer.renderSelectedCells(model,
+                                     context,
+                                     rendererHelper).execute(rc);
+
+        verify(renderer, never()).renderSelectedRange(any(GridData.class),
```

Comment:
Well, now I am confused, please, what does the `BaseGridRenderer` for `SelectionLayer` does, when this test confirm it doesn't render selected cells.

Reply:
OK, certain things are not rendered to the ```SelectionLayer``` only column backgrounds are rendered to that layer. Lienzo uses the ```SelectionLayer``` to handle events. When an event occurs Lienzo looks up a colour at the event co-ordinate on the SL from which it works out what ```Shape``` the event occurred on. To improve performance (not by this PR, but in the original version) a lot of "decoration" was not rendered to the SL (cell selection highlights, grid lines, grid border etc) as we do not listen for events on these decorations. So the test checks that commands for these "decorations" to not result in calls to the renderer (that actually adds things to the layer). Any clearer? ;-)